runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -0,0 +1,642 @@
+     * @param <P>   The type returned by the projection function.
+     */
+    public interface IndexSpecification<K, V, I, P> {
+        IndexFunction<K, V, I> getIndexFunction();
+        ProjectionFunction<K, V, I, P> getProjectionFunction();
+    }
+
+    /** An index specification enumeration which has no index specifications.
+     *  Using this enumeration effectively disables secondary indexes.
+     */
+    enum NoSecondaryIndex implements IndexSpecification {
+        ;
+
+        @Override
+        public IndexFunction getIndexFunction() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public ProjectionFunction getProjectionFunction() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    /** The "main" map which contains the primary key-value mappings. */
+    protected final Map<K,V> mainMap = new HashMap<>();
+
+    protected Set<F> indexFunctions;
+
+    protected Map<F, Multimap<I, Map.Entry<K,V>>> indexMap;
+
+    /** Generate a table with the given set of indexes. */
+    public CorfuTable(Class<F> indexFunctionEnumClass) {
+        indexerClass = indexFunctionEnumClass;
+        indexMap = new HashMap<>();
+        indexFunctions = EnumSet.allOf(indexFunctionEnumClass);
+        indexFunctions.forEach(f -> indexMap.put(f, ArrayListMultimap.create()));
+    }
+
+    /** Default constructor. Generates a table without any secondary indexes. */
+    public CorfuTable() {
+        indexFunctions = Collections.emptySet();
+        indexMap = Collections.emptyMap();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public int size() {
+        return mainMap.size();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public boolean isEmpty() {
+        return mainMap.isEmpty();
+    }
+
+    /** Return whether this table has secondary indexes or not.
+     *
+     * @return  True, if secondary indexes are present. False otherwise.
+     */
+    @Accessor
+    public boolean hasSecondaryIndices() {
+        return !indexFunctions.isEmpty();
+    }
+
+    @Getter
+    public Class indexerClass;
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public boolean containsKey(@ConflictParameter Object key) {
+        return mainMap.containsKey(key);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public boolean containsValue(Object value) {
+        return mainMap.containsValue(value);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public V get(@ConflictParameter Object key) {
+        return mainMap.get(key);
+    }
+
+
+
+    /** Get a mapping using the specified index function.
+     *
+     * @param indexFunction The index function to use.
+     * @param index         The index
+     * @return
+     */
+    @SuppressWarnings("unchecked")
+    @DontInstrument
+    public @Nonnull Collection<Object> getByIndex(@Nonnull F indexFunction, I index) {
```

Comment:
What do you mean not expose the "index"? ... That doesn't make any sense, the user needs to specify a secondary index